### PR TITLE
Move error message and hint out of labels and legends

### PR DESCRIPTION
**Pull request by alex-ju**
This PR moves error messages and hints out of legends and labels, instead associating them with the input using `aria-describedby`.

https://govuk-frontend-review-pr-681.herokuapp.com/

https://trello.com/c/6JHtnTfC

Co-authored with @36degrees.

## Background

For question pages that follow the 'one-thing-per-page' format, the question being asked functions as both the label for the question and the H1 for the page.

screen shot 2018-05-03 at 13 09 09

This was previously achieved by putting the question in the H1 but duplicating it within a _visually hidden label or legend_. This was functional, but resulted in the question being read twice by screen readers. Anika [did a lot of research into this and made changes to Elements to remove the duplication](https://github.com/alphagov/govuk_elements/pull/507).

The way this currently works is:

- putting the `<h1>` _inside_ the legend of a fieldset (`<legend><h1>Question</h1></legend>`); or
- wrapping the `<label>` in an `<h1>` (`<h1><label>Question</label></h1>`)

The reason that the nesting differs is that fieldsets allow '[phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content) and headings' as permitted content (since https://github.com/w3c/html/issues/724) but labels only allow for 'phrasing content', so a heading inside a label is technically invalid.

There are a few issues that we would like to address to make this approach work correctly in GOV.UK Frontend:

- If the form field has associated hint text and an error message then these become _part of the h1_ for the page as well.
- because the hint and error message are part of the label, clicking them focusses the input. This confuses things if the hint or error message contain a link – for example to additional guidance or resources.
- it is difficult to make the relationship between the question and the hint and error message consistent with the `fieldset`, because of the inconsistency in the markup.

Having explored a number of options (including [questioning whether the HTML spec should allow headings inside labels](https://github.com/w3c/html/issues/1270)), we've settled on moving the hint and error message outside of the label / legend and associating them with `aria-describedby` as [has been previously suggested](https://github.com/alphagov/govuk_elements/issues/574) - it resolves the first two issues and makes it considerably easier to tackle the third.

Unfortunately Voiceover on macOS will not announce the hint or error message when focussing the field. This is not ideal, but every approach (including the current approach of nesting them inside the label) had issues in at least one combination of technologies, so we have opted to go with the approach that seemed the most 'semantically correct', which is to treat the error message and hints as part of the 'description' of the field as per [the ARIA spec](https://www.w3.org/TR/wai-aria-1.1/#aria-describedby):

> The aria-labelledby attribute is similar to the aria-describedby in that both reference other elements to calculate a text alternative, but a label should be concise, where a description is intended to provide more verbose information.

~~We will file an issue with Apple if we can not find it already recorded as a bug.~~

[We have filed a bug report with Apple for this issue](https://bugs.webkit.org/show_bug.cgi?id=185246).

We considered using different approaches for different scenarios (for example, only using `aria-describedby` if the `legend` or `label` sits inside a heading, or using different implementations depending on whether we are dealing with a `legend` or a `label`). However, we concluded that there were a number of benefits to having a consistent approach, as: 

- assistive technologies should announce the field(s) in a consistent manner regardless of the field type or page structure
- it reduces the likelihood of service teams introducing errors when implementing
- it reduces the complexity of the logic required in the macros for the different components

### Detailed changes

- moves errorMessage component out of label (for every component that uses a label)
- moves errorMessage component out of legend  (for every component that uses a fieldset)
- adds a Hint component (similar with the errorMessage component; up for discussions if this component should be treated separately in the Design System)
- moves hint out of label (for every component that uses a label)
- moves hint out of legend  (for every component that uses a fieldset)
- updates SCSS dependencies for each updated component
- update tests to reflect the changes mentioned above
- add a dependency on `/tools/exports` to `helpers/clearfix` in order to pass the _Individual components should compile individual SCSS files without throwing exceptions_

### To be done in future PRs:
- update the styling (one obvious issue is the bottom spacing on the legend, which visually separates it from the hint and the error message)
  screen shot 2018-05-03 at 11 01 15
- handle the headings in labels and legends

#### Input with `<label>`

❗️When the `<label>` is nested inside a heading, the hint and error message also become part of the heading.

❗️It is difficult to make the relationship between the question and the hint and error message consistent with the `fieldset`, because of the inconsistency in the markup.

<details>
<summary>✅ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> "National insurance number It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number edit text
> 
> You are currently on a text field, inside of HTML content. To enter text in this field, type. To exit this web area, press Control-Option-Shift-Up Arrow."
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 11)</summary>  

Voiceover announces:

> National insurance number It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, text field.
> 
> Double tap to edit.
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 10)</summary>  

Voiceover announces:

> National insurance number It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, text field.
> 
> Double tap to edit.
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 9)</summary>  

Voiceover announces:

> National insurance number It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, text field.
>
> Double tap to edit."
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> National insurance number. It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number. Edit. Blank.

</details>

<details>
<summary>✅ JAWS18 + IE11</summary>

JAWS announces:

> National insurance number.
> It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’.
> Enter your national insurance number.
> Edit. Type in text.
</details>

#### Fieldset with `<legend>`

Very strange issue in JAWS18 where the hint and error message are not announced. Other than that, this works OK – because the heading is inside the legend (unlike label) the hint and error message can sit outside of the page heading. (Update: This appears to have been fixed in JAWS 2018)

<details>
<summary>✅ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> "Yes, radio button, 1 of 2, Have you changed your name? This includes changing your last name or spelling your name differently. Choose yes or no.
>
> You are currently on a radio button, 1 of 2, inside of a group. To select this option…"
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 11)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name? This includes changing your last name or spelling your name differently. Choose yes or no.

Then, for each input:

> Yes, radio button, 1 of 2.
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 10)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name? This includes changing your last name or spelling your name differently. Choose yes or no.

Then, for each input:

> Yes, radio button, 1 of 2.
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 9)</summary>  

For each input, Voiceover announces:

> Yes, radio button, 1 of 2. This includes changing your last name or spelling your name differently. Choose yes or no.

iOS 9 does not seem to announce `<fieldset>` legends for inputs.
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> Have you changed your name?
> This includes changing your last name or spelling your name differently.
> Choose yes or no. Grouping.
> Yes, radio button, not checked, 1 of 2.

</details>

<details>
<summary>❗️ JAWS18 + IE11</summary>

JAWS announces:

> Have you changed your name?
> Yes, radio button, not checked.

Hint and error message are not announced despite being part of the legend. Absolutely no idea what's going on here. Tested this with the current implementation in Elements and the same thing happens.
</details>

### Associating error and hint with `aria-describedby`

Seems like the 'correct' implementation semantically. Some slight quirks in the way this is announced, e.g. JAWS puts the description between "Edit" and "Type in text" which seems disjointed.

For legends, we should associate the hint and error message with the fieldset, not the individual inputs.

#### Input with `<label>`

<details>
<summary>✅ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> National insurance number, edit text
>  
> It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, You are currently on a text field, inside of HTML content. To enter text in this field, type. To exit this web area, press Control-Option-Shift-Up Arrow.
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 11)</summary>  

Voiceover announces:

> National insurance number, text field
> 
> It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, double tap to edit
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 10)</summary>  

Voiceover announces:

> National insurance number, text field
> 
> It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, double tap to edit
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 9)</summary>  

Voiceover announces:

> National insurance number, text field
> 
> It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number, double tap to edit
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> National insurance number. Edit. It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number. Blank.

</details>

<details>
<summary>✅ JAWS18 + IE11</summary>

JAWS announces:

> National insurance number. Edit. Use JAWS key + alt + r to read descriptive text.
It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’.
Enter your national insurance number.
Type in text.

(includes JAWS commands)
</details>

#### Fieldset with `<legend>`, `aria-describedby` on each `<input>`

<details>
<summary>✅ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> Yes, radio button, 1 of 2, Have you changed your name?

This includes changing your last name or spelling your name differently. Choose yes or no. You are currently on a radio button, 1 of 2, inside of a group. To select this option…
</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 11)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name?

Then, for each input:

> Yes, radio button, 1 of 2. This includes changing your last name or spelling your name differently. Choose yes or no.

Hint and error message are always announced when focussing the input, but the legend is only announced when entering the fieldset. This behaviour may be confusing as the hint is unlikely to make sense without the question being announced. It's also overly verbose.

</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 10)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name?

Then, for each input:

> Yes, radio button, 1 of 2. This includes changing your last name or spelling your name differently. Choose yes or no.

Hint and error message are always announced when focussing the input, but the legend is only announced when entering the fieldset. This behaviour may be confusing as the hint is unlikely to make sense without the question being announced. It's also overly verbose.

</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 9)</summary>  

For each input, Voiceover announces:

> Yes, radio button, 1 of 2. This includes changing your last name or spelling your name differently. Choose yes or no.

iOS 9 does not seem to announce `<fieldset>` legends for inputs.
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> "Have you changed your name? Grouping.
> Yes, radio button, not checked.
> This includes changing your last name or spelling your name differently. Choose yes or number. 1 of 2."

There is a minor bug here ([reported](https://github.com/nvaccess/nvda/issues/8255)) where no is announced as number, but it is a side effect of the ordering.

</details>

<details>
<summary>✅ JAWS18 + IE11</summary>

JAWS announces:

> "Have you changed your name?
Yes, radio button, not checked. This includes changing your last name or spelling your name differently.
Choose yes or no."
</details>

#### Fieldset with `<legend>`, `aria-describedby` on the `<fieldset>`

<details>
<summary>❗️ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> Yes, radio button, 1 of 2, Have you changed your name?
> 
> You are currently on a radio button, 1 of 2, inside of a group. To select this option…

The error message and hint are not announced.
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 11)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name?
>
> This includes changing your last name or spelling your name differently. Choose yes or no.""

Then, for each input:

> Yes, radio button, 1 of 2.

</details>

<details>
<summary>✅ Voiceover + Safari (iOS 10)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name?
>
> This includes changing your last name or spelling your name differently. Choose yes or no.""

Then, for each input:

> Yes, radio button, 1 of 2.

</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 9)</summary>  

For each input, Voiceover announces:

> Yes, radio button, unticked, 1 of 2.

iOS 9 does not seem to announce `<fieldset>` legends for inputs.
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> "Have you changed your name? Grouping.
> Yes, radio button, not checked.
> This includes changing your last name or spelling your name differently. Choose yes or number. 1 of 2."

There is a minor bug here where no is announced as number, but it is a side effect of the ordering.

</details>

<details>
<summary>✅ JAWS18 + IE11</summary>

JAWS announces:

> "Have you changed your name?
Yes, radio button, not checked. This includes changing your last name or spelling your name differently.
Choose yes or no."
</details>

### Associating error and hint with `aria-labelledby`

According to the ARIA spec, a label should be concise, where a description is intended to provide more verbose information. This suggests that aria-labelledby is not the right implementation for the hint and error message, as they do not label the field and cannot really be considered 'concise'.

#### Input with `<label>`

<details>
<summary>✅ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> "National insurance number It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number edit text

> You are currently on a text field, inside of HTML content. To enter text in this field, type. To exit this web area, press Control-Option-Shift-Up Arrow."
</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 11)</summary>  

Voiceover announces:

> National insurance number, text field, double tap to edit.

A [bug in iOS](https://bugs.webkit.org/show_bug.cgi?id=172464) means that if a label is associated with a field, that label will override any aria defined label.
</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 10)</summary>  

Voiceover announces:

> National insurance number, text field, double tap to edit.

A [bug in iOS](https://bugs.webkit.org/show_bug.cgi?id=172464) means that if a label is associated with a field, that label will override any aria defined label.
</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 9)</summary>  

Voiceover announces:

> National insurance number, text field, double tap to edit.

A [bug in iOS](https://bugs.webkit.org/show_bug.cgi?id=172464) means that if a label is associated with a field, that label will override any aria defined label.
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> National insurance number. It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’. Enter your national insurance number. Edit. Blank.

</details>

<details>
<summary>✅ JAWS18 + IE11</summary>

JAWS announces:

> "National insurance number.
> It’s on your National Insurance card, benefit letter, payslip or P60. For example, ‘QQ 12 34 56 C’.
> Enter your national insurance number.
> Edit. Type in text."

(includes JAWS commands)
</details>

#### Fieldset with `<legend>`

<details>
<summary>✅ Voiceover + Safari (Mac OS X El Capitan)</summary>

Voiceover announces:

> "Yes, radio button, 1 of 2, Have you changed your name? This includes changing your last name or spelling your name differently. Choose yes or no.
> 
> You are currently on a radio button, 1 of 2, inside of a group. To select this option…"
</details>

<details>
<summary>✅ Voiceover + Safari (iOS 11)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name? This includes changing your last name or spelling your name differently. Choose yes or no.

Then, for each input:

> Yes, radio button, 1 of 2.

</details>

<details>
<summary>✅️ Voiceover + Safari (iOS 10)</summary>  

When entering the fieldset (e.g. navigating backwards through the document)", Voiceover announces:

> Have you changed your name? This includes changing your last name or spelling your name differently. Choose yes or no.

Then, for each input:

> Yes, radio button, 1 of 2.

</details>

<details>
<summary>❗️ Voiceover + Safari (iOS 9)</summary>  

For each input, Voiceover announces:

> Yes, radio button, 1 of 2. This includes changing your last name or spelling your name differently. Choose yes or no.

iOS 9 does not seem to announce `<fieldset>` legends for inputs.
</details>

<details>
<summary>✅ NVDA + Firefox</summary>

NVDA announces:

> Have you changed your name?
> This includes changing your last name or spelling your name differently.
> Choose yes or no. Grouping.
> Yes, radio button, not checked, 1 of 2.

</details>

<details>
<summary>✅ JAWS18 + IE11</summary>

JAWS announces:

> "Have you changed your name?
> This includes changing your last name or spelling your name differently.
> Choose yes or no.
> Have you changed your name?
> Yes, radio button, not checked."
</details>

### CODE REVIEW

File: src/hint/hint.njk [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{% from "hint/macro.njk" import govukHint %}
+
+{{ govukHint({
+  text: "It’s on your National Insurance card, benefit letter, payslip or P60.
+    For example, ‘QQ 12 34 56 C’.",
+  id: "hint-id",
+  name: "hint-name"
```
Comment (nickcolley) [inline src/hint/hint.njk]: Is `name` used?

Comment (alex-ju) [reply src/hint/hint.njk]: nope. good catch 👍

File: src/checkboxes/template.njk [hunk 2]
```diff
@@ -1,6 +1,16 @@
+{% from "error-message/macro.njk" import govukErrorMessage -%}
 {% from "fieldset/macro.njk" import govukFieldset %}
+{% from "hint/macro.njk" import govukHint %}
 {% from "label/macro.njk" import govukLabel %}
 
+{#- If an id 'prefix' is not passed, fall back to using the name attribute
+   instead. We need this for error messages and hints as well -#}
+{% set idPrefix = params.idPrefix if params.idPrefix else params.name %}
```
[the remark this answers is not in the document]
Comment (igloosi) [reply src/checkboxes/template.njk]: hmm, don't requires it to be. 
if you had two sets of checkbox groups on the page (nationality, religion) then data for all of them would be submitted as `nationality=<value>` even if the second set was about religion. We can't solve this (the same would occur with idPrefix)

**Comment by selfthinker** [comment]
Great and thorough work, well done!

One thing I already mentioned to Oli... There are generally three cases to test not just two.
1. Label for input
2. fieldset and legend for radio buttons or checkboxes
3. fieldset and legend for multiple inputs, e.g. date of birth

I believe you haven't tested the latter. It might make a difference or it might not. But I think it makes sense to check that as well, especially regarding where the `aria-describedby` should go.

**Comment by alex-ju** [comment]
Good call @selfthinker! In my naivety, I thought any input, regardless its type, will be treated similarly by the ATs. @36degrees and I did a round of tests today on the date input example, and it turns out it's not the case with JAWS - which for the date input example doesn't announce the hint and error message when landing on the input fields, despite the fact that they are being referred in the _describedby_ attribute. We'll think about workarounds for JAWS to be able to pick them up.

**Comment by alex-ju** [comment]
@nickcolley tests to make sure aria-describedby contains both the hint and the error message and in the right order might be a good addition.
Also accessibility criteria should be added, but I'm not sure will get into this commit until we define generic cirterias.
I would thumb up your comment, but I can't since it's a review.

**Comment by nickcolley** [comment]
There's nothing stopping us from adding additional information right now.  https://github.com/alphagov/govuk-frontend/search?utf8=%E2%9C%93&q=accessibilityCriteria&type=

Would be good to have this since this is fairly nuanced.

**Comment by 36degrees** [comment]
We haven't been able to make Voiceover on macOS read the error message and hint when focussing the field. It doesn't read the description for the fieldset. We've [filed a bug with Apple](https://bugs.webkit.org/show_bug.cgi?id=185246) as we don't think this is correct.

However, the error message is available as part of the error summary component (which acts as an alert, so should always be announced to the user), and both the hint and error message _can_ be accessed by 'going out of' the input (ctrl-option-shift + up-arrow) which focuses the fieldset.

This change also gives us a number of benefits. It:

- fixes an issue where the error message and hint are not announced in JAWS 18, which we understand to be considerably more popular according to the [WebAIM screen reader survey](https://webaim.org/projects/screenreadersurvey7/#primary) [2017] and our [own research](https://accessibility.blog.gov.uk/2016/11/01/results-of-the-2016-gov-uk-assistive-technology-survey/) [2016].
- removes the hint and error message from the page heading.
- allowing us to better define the relationship between the question, hint and error message.

With that in mind, we're going to merge this.

**Pull request merged**